test/common/config/delta_subscription_impl_test.cc

```diff
@@ -77,14 +78,103 @@ TEST_F(DeltaSubscriptionImplTest, ResourceGoneLeadsToBlankInitialVersion) {
 
   // ...but our own map should remember our interest. In particular, losing interest in all 3 should
   // cause their names to appear in the resource_names_unsubscribe field of a DeltaDiscoveryRequest.
-  subscription_->resume(); // now we do want the request to actually get sendMessage()'d.
-  EXPECT_CALL(async_stream_, sendMessage(_, _)).WillOnce([](const Protobuf::Message& msg, bool) {
-    auto sent_request = static_cast<const envoy::api::v2::DeltaDiscoveryRequest*>(&msg);
-    EXPECT_THAT(sent_request->resource_names_subscribe(), UnorderedElementsAre("name4"));
-    EXPECT_THAT(sent_request->resource_names_unsubscribe(),
-                UnorderedElementsAre("name1", "name2", "name3"));
-  });
-  subscription_->subscribe({"name4"}); // (implies "we no longer care about name1,2,3")
+  subscription_->resume(); // we do want the final subscribe() to do a sendMessage().
+  expectSendMessage({"name4"}, {"name1", "name2", "name3"}, Grpc::Status::GrpcStatus::Ok, "");
+  subscription_->updateResources({"name4"}); // (implies "we no longer care about name1,2,3")
+}
+
+// Delta xDS reliably queues up and sends all discovery requests, even in situations where it isn't
+// strictly necessary. E.g.: if you subscribe but then unsubscribe to a given resource, all before a
+// request was able to be sent, two requests will be sent. The following tests test various cases of
+// this reliability. TODO TODO REMOVE PROBABLY
+//
+// If Envoy decided it wasn't interested in a resource and then (before a request was sent) decided
+// it was again, for all we know, it dropped that resource in between and needs to retrieve it
+// again. So, we *should* send a request "re-"subscribing. This means that the server needs to
+// interpret the resource_names_subscribe field as "send these resources even if you think Envoy
+// already has them".
+TEST_F(DeltaSubscriptionImplTest, RemoveThenAdd) {
+  startSubscription({"name1", "name2", "name3"});
+  subscription_->pause(); // Pause because we're testing multiple updates in between request sends.
+  subscription_->updateResources({"name1", "name2"});
+  subscription_->updateResources({"name1", "name2", "name3"});
+  InSequence s;
+  expectSendMessage({"name3"}, {}, Grpc::Status::GrpcStatus::Ok, "");
+  expectSendMessage({}, {}, Grpc::Status::GrpcStatus::Ok, ""); // no-op due to the second update
+  subscription_->resume();
+}
+
+// Due to how our implementation provides the required behavior tested in RemoveThenAdd, the
+// add-then-remove case *also* causes the resource to be referred to in the request (as an
+// unsubscribe).
+// Unlike the remove-then-add case, this one really is unnecessary, and ideally we would have
+// the request simply not include any mention of the resource. Oh well.
```

Comment:
Should this one be a TODO?

Reply:
I don't think the code complexity would be worth it, like even if the labor to make the change was free. It would only be avoiding an extraneous-but-not-harmful message in a corner case.

Reply:
I think it's not worth the code complexity, like even if the labor was free. The benefit would just be a small efficiency gain in a pretty corner-y corner case.